src/libs/auth.ts

```diff
@@ -109,59 +149,146 @@ const getSigninArgs = (includeBillingScope) => {
 };
 
 export const signIn = async (includeBillingScope = false): Promise<User> => {
-  const args = getSigninArgs(includeBillingScope);
-  const user: User = await getAuthInstance().signinPopup(args);
+  // we should handle if we get back null or false here (if loading the authTokenFails)
+  const user: User | null | boolean = await loadAuthToken(includeBillingScope, true);
+  if (user instanceof User) {
+    const sessionId = uuid();
+    const sessionStartTime: number = Date.now();
+    authStore.update((state) => ({
+      ...state,
+      hasGcpBillingScopeThroughB2C: includeBillingScope,
+      sessionId,
+      sessionStartTime,
+    }));
+    Ajax().Metrics.captureEvent(Events.user.login, {
+      sessionStartTime: Utils.makeCompleteDate(sessionStartTime),
+    });
+    return user;
+  }
+  throw new Error('Auth token failed to load when signing in');
+};
 
-  const sessionId = uuid();
-  const sessionStartTime: number = Date.now();
-  const userJWT: string = user.id_token!;
-  const decodedJWT: JwtPayload = jwtDecode<JwtPayload>(userJWT);
-  const authTokenCreatedAt: number = (decodedJWT as any).auth_time; // time in seconds when authorization token was created
-  const authTokenExpiresAt: number = user.expires_at!; // time in seconds when authorization token expires, as given by the oidc client
-  const jwtExpiresAt: number = (decodedJWT as any).exp; // time in seconds when the JWT expires, after which the JWT should not be read from
+export const loadAuthToken = async (includeBillingScope = false, popUp = false): Promise<User | null | boolean> => {
+  // here we should update the auth token id
+  // might want to take refresh token and encrypt it to compare to a saved encrypted token
+  // to see if they are the same => track refresh update
+  // track lifespan of current authToken
+  // need to send information about the last authToken
+  const oldAuthTokenMetadata = authStore.get().authTokenMetadata; // this can be null (first token), so cover that case
 
   authStore.update((state) => ({
     ...state,
-    hasGcpBillingScopeThroughB2C: includeBillingScope,
-    sessionId,
-    sessionStartTime,
     authTokenMetadata: {
-      createdAt: authTokenCreatedAt,
-      expiresAt: authTokenExpiresAt,
+      ...oldAuthTokenMetadata,
+      totalAuthTokenLoadAttemptsThisSession:
+        authStore.get().authTokenMetadata.totalAuthTokenLoadAttemptsThisSession + 1,
     },
   }));
-  Ajax().Metrics.captureEvent(Events.userLogin, {
-    authProvider: user.profile.idp,
-    sessionStartTime: Utils.makeCompleteDate(sessionStartTime),
-    authTokenCreatedAt: Utils.formatTimestampInSeconds(authTokenCreatedAt),
-    authTokenExpiresAt: Utils.formatTimestampInSeconds(authTokenExpiresAt),
-    jwtExpiresAt: Utils.formatTimestampInSeconds(jwtExpiresAt),
-  });
-  return user;
+
+  const reloadedAuthTokenState: User | boolean | null = popUp
```

Comment:
Multiple possible return types here seems like a code smell. For example of how this makes the code more difficult to understand, the type signature suggests that `reloadedAuthTokenState` may be true, but the if statement below treats both true or false as an error.

I would suggest letting the `tryLoadAuthToken...` methods throw instead of returning `false` when there's an error.

Reply:
I agree that it is a smell. We are thinking to make the `User | boolean | null` union type into its own specific type with the states: `Success => User, Error, RefreshTokenExpired => null`.

Reply:
Ok. Not sure exactly what you're thinking, but something like LoadedState might be good pattern to follow.

https://github.com/DataBiosphere/terra-ui/blob/dev/packages/core-utils/src/type-utils/LoadedState.ts

Reply:
I added a union type, but I did notice that the case matching was not so straightforward. Maybe I am doing it in a suboptimal way.